build: automate macOS package notarization (#1406)
radicle-dev/radicle-upstream: DEVELOPMENT.md

7. Commit changes to `Dockerfile` and `pipeline.yaml`. Pushing the changes will
   create a new branch and build the updated image.

## Setting up Apple notarization

> To [allow macOS Gatekeeper recognise our Upstream packages as genuine][so],

Reviewer: ```suggestion
To [allow macOS Gatekeeper to recognise our Upstream packages as genuine][so],
```

which allows the user to install and open Upstream without unnecessary
[security warnings][sw], we have to [sign and notarize][sn] our macOS packages.

Prerequisites:

>   - a paid Apple developer account registered to our company

Reviewer 2: ```suggestion
  - a paid Apple developer account registered to Monadic
```

>   - an Apple ID token for allowing the notarization script to run in behalf of

Reviewer 2: ```suggestion
  - an Apple ID token for allowing the notarization script to run on behalf of
```

>     our developer acocunt

Reviewer: ```suggestion
    our developer account
```

    - [Account Manage][ma] -> APP-SPECIFIC PASSWORDS -> Generate password…
  - a valid "Developer ID Application" certificate
    - [Certificates Add][ca] -> Developer ID Application

>       **Note:** this can only be created via the company account holder account

Reviewer 2: ```suggestion
      **Note:** this can only be created via the company account holder
```

instructions.

Once the release PR branch is merged into master a build will be triggered on

> Buildkite, this will build Upstream for both Linux and macOS (unsigned).

Reviewer 3: Does it make sense to rename this section to “Preparing a release”? Just “Release” makes it sound like it is end-to-end.

Author: https://github.com/radicle-dev/radicle-upstream/pull/1406/commits/354b7e5bae9dd34e141c560ba92e32672b3cdd10

<details>

> <summary>This is what a typical release looks like</summary>

Reviewer 3: I was confused by this and thought it would show me an example of the release artifacts.
```suggestion
<summary>Commands to prepare a release</summary>
```

instructions.

> Once the release PR branch is merged into master a build will be triggered on

Reviewer: ```suggestion
Once the release PR branch is merged into master, a build will be triggered on
```

on a developer's macOS machine.

If you haven't already, please set up an Apple developer account and signing
certificates, see [Setting up Apple notarization][an] for more details.

To build, sign and notarize a macOS dmg package of the latest version run the
following commands:

```
cd radicle-upstream
git checkout v0.0.11

CSC_NAME="Monadic GmbH (XXXXXXXXXX)" \
APPLE_ID="XXXXXXX@monadic.xyz" \
APPLE_ID_PASSWORD="XXXX-XXXX-XXXX-XXXX" \
yarn dist
```

Where:
  - `CSC_NAME` is the name of the signing certificate
  - `APPLE_ID` is the developer's Apple developer ID
  - `APPLE_ID_PASSWORD` is an app specific token generated from your Apple ID
    in the developer portal

**Note**: building a release might take a while, especially the notarization
step, because it has to upload the final package to the Apple notarization
server. Make sure you're on a stable internet connection.

Once the package is built, signed and notarized you should upload it to our
Google Cloud Platform under
[`builds.radicle.xyz/releases/radicle-upstream/v0.1.4`][gp].

You should also copy the Linux `.AppImage` and `.snap` packages to the same

> location. You can get them from the [CI build][ar] on Buildkite.

Reviewer 3: We should provide command snippets for these steps and define the filenames in the GCS bucket.

Author: Was added in: https://github.com/radicle-dev/radicle-upstream/pull/1411.

After all the packages are uploaded, update the links to those binaries on the

> [radicle.xyz download][rd] and [docs.radicle.xyz/docs/getting-started][gs]

Reviewer 3: Links to the source code on Github would be nice. Ideally it would just be one location where we need to change this.

Author: https://github.com/radicle-dev/radicle-upstream/issues/1410

someone from the team goes through the checklist and manually tests the app,
afterwards the team can evaluate whether the release is up to our standards.

**Title:** `QA: v0.0.11 macOS`\
**Body** [QA.md][qa]

> **Title:** `QA: v0.0.11 Linux`\

Reviewer 2: What's with the version change?

Author: To make it consistent with the rest of the instructions.

**Body:** [QA.md][qa]

## Publishing a release

Once a release has passed QA, it is ready to be published. This involves a
couple of manual steps since the macOS release has to be signed and notarized
on a developer's macOS machine.

If you haven't already, please set up an Apple developer account and signing
certificates, see [Setting up Apple notarization][an] for more details.

To build, sign and notarize a macOS dmg package of the latest version run the
following commands:

```
cd radicle-upstream
git checkout v0.0.11

CSC_NAME="Monadic GmbH (XXXXXXXXXX)" \
APPLE_ID="XXXXXXX@monadic.xyz" \
APPLE_ID_PASSWORD="XXXX-XXXX-XXXX-XXXX" \
yarn dist
```

Where:
  - `CSC_NAME` is the name of the signing certificate
  - `APPLE_ID` is the developer's Apple developer ID
  - `APPLE_ID_PASSWORD` is an app specific token generated from your Apple ID
    in the developer portal

**Note**: building a release might take a while, especially the notarization
step, because it has to upload the final package to the Apple notarization
server. Make sure you're on a stable internet connection.

Once the package is built, signed and notarized you should upload it to our
Google Cloud Platform under

> [`builds.radicle.xyz/releases/radicle-upstream/v0.1.4`][gp].

Reviewer: Above, we mention `v0.0.11` but here we mention `v0.1.4` :pencil2: 
```suggestion
[`builds.radicle.xyz/releases/radicle-upstream/v0.0.11`][gp].
```
